File: pronacej-front/src/app/modules/administracion/services/seguimientoSocial.service.ts
import { Injectable } from '@angular/core';
import { BodyEncriptado } from 'app/core/model/both/bodyEncriptado.model';
import { DocumentoDTO } from 'app/core/model/both/DocumentoDTO.model';
import { SeguimientoEducativoDTO } from 'app/core/model/both/ia/seguimientoEducativoDTO.model';
import { SeguimientoSocialDTO } from 'app/core/model/both/ia/SeguimientoSocialDTO.model';
import { PaginacionRequest } from 'app/core/model/request/PaginacionRequest.model';
import { PaginacionResponse } from 'app/core/model/response/PaginacionResponse.model';
import { RespuestaPorDefecto } from 'app/core/model/response/RespuestaPorDefecto.model';
import { BackendService } from 'app/core/services/backend.service';
import { Observable, Subscriber } from 'rxjs';

@Injectable({
    providedIn: 'root',
})
export class SeguimientoSocialService {
    private path = '/seguimiento-social';

    constructor(private backendService: BackendService) {}

    /**
     * Obtiene los seguimientos sociales disponibles para el sistema de manera paginada
     *
     * @param paginacionRequest PaginacionRequest objeto con los parámetros de paginación
     * @param nemonicoMenu string nemonico del menú del sistema
     *
     * @returns Observable<RespuestaPorDefecto<PaginacionResponse<SeguimientoSocialDTO>>>
     */
    obtenerSeguimientosSocialesPaginado(
        paginacionRequest: PaginacionRequest,
        nemonicoMenu = ''
    ): Observable<
        RespuestaPorDefecto<PaginacionResponse<SeguimientoSocialDTO>>
    > {
        let endPoint = this.path + '/obtenerSeguimientosSocialesPaginado';
        return this.backendService.postFinal(
            endPoint,
            paginacionRequest,
            nemonicoMenu
        );
    }

    /**
     * Crea un seguimiento social en el sistema con los datos enviados en el request
     *
     * @param seguimientoSocialDTO SeguimientoSocialDTO datos del seguimiento social a crear
     * @param nemonicoMenu string nemonico del menú del sistema
     *
     * @returns Observable<RespuestaPorDefecto<SeguimientoSocialDTO>>
     */
    crearSeguimientoSocial(
        seguimientoSocialDTO: SeguimientoSocialDTO,
        nemonicoMenu: string = ''
    ): Observable<RespuestaPorDefecto<SeguimientoSocialDTO>> {
        let endPoint = this.path + '/crearSeguimientoSocial';
        return this.backendService.postFinal(
            endPoint,
            seguimientoSocialDTO,
            nemonicoMenu
        );
    }

    /**
     * Elimina un seguimiento social en el sistema
     *
     * @param seguimientoSocialDTO SeguimientoSocialDTO datos del seguimiento social a eliminar
     * @param nemonicoMenu string nemonico del menú del sistema
     *
     * @returns Observable<RespuestaPorDefecto<boolean>>
     */
    eliminarSeguimientoSocial(
        seguimientoSocialDTO: SeguimientoSocialDTO,
        nemonicoMenu = ''
    ): Observable<RespuestaPorDefecto<boolean>> {
        let endPoint = this.path + '/eliminarSeguimientoSocial';
        return this.backendService.postFinal(
            endPoint,
            seguimientoSocialDTO,
            nemonicoMenu
        );
    }

    /**
     * Sube un informe firmado
     *
     * @param encabezadoDTO InformeDTO
     * @param archivo File
     * @param nemonicoMenu string nemonico menu
     *
     * @return Observable<RespuestaPorDefecto<Boolean>>
     */
    subirDocumentos(
        seguimientoDTO: SeguimientoEducativoDTO,
        files: File[],
        nemonicoMenu: string
    ): Observable<RespuestaPorDefecto<Boolean>> {
        let endPoint = this.path + '/subirDocumentos';

        let formData = new FormData();
        if (files != null) {
            for (let file of files) {
                formData.append('documentos', file);
            }
        }

        return new Observable(
            (subs: Subscriber<RespuestaPorDefecto<Boolean>>) => {
                this.backendService
                    .crearBodyEncriptado(seguimientoDTO)
                    .then((bodyEncriptado) => {
                        formData.append('body', JSON.stringify(bodyEncriptado));
                        this.backendService
                            .postFormDataBodyEncriptado2(
                                endPoint,
                                formData,
                                nemonicoMenu
                            )
                            .subscribe({
                                next: async (body: BodyEncriptado) => {
                                    let resp =
                                        await this.backendService.desencriptarBdyEncriptado<
                                            RespuestaPorDefecto<Boolean>
                                        >(body);
                                    subs.next(resp);
                                    subs.complete();
                                },
                                error: (error: any) => {
                                    subs.error(error);
                                    subs.complete();
                                },
                            });
                    })
                    .catch((error: any) => {
                        subs.error(error);
                        subs.complete();
                    });
            }
        );
    }

    obtenerDocumentos(
        request: PaginacionRequest,
        nemonicoMenu: string
    ): Observable<RespuestaPorDefecto<PaginacionResponse<DocumentoDTO>>> {
        let endPoint = this.path + '/obtenerDocumentos';
        return this.backendService.postFinal(endPoint, request, nemonicoMenu);
    }

    /**
     * Verifica y maneja los errores generales del servicio
     *
     * @param error any error a verificar
     * @param mostrarError boolean indica si se debe mostrar el error
     * @returns any
     */
    async checkError(error: any, mostrarError = true): Promise<string> {
        return await this.backendService.checkError(error, mostrarError);
    }
}

File: pronacej-backend-seguridad/src/main/java/net/latinus/sistema/integral/gestion/seguridad/service/IA/SeguimientoSocialService.java
package net.latinus.sistema.integral.gestion.seguridad.service.IA;

import jakarta.servlet.http.HttpServletRequest;
import net.latinus.sistema.integral.gestion.seguridad.model.both.BodyEncriptado;
import net.latinus.sistema.integral.gestion.seguridad.model.both.DocumentoDTO;
import net.latinus.sistema.integral.gestion.seguridad.model.both.SeguimientoSocialDTO;
import net.latinus.sistema.integral.gestion.seguridad.model.response.PaginacionResponse;
import net.latinus.sistema.integral.gestion.seguridad.model.response.RespuestaPorDefectoAuditoria;
import org.springframework.web.multipart.MultipartFile;

public interface SeguimientoSocialService {

    /**
     * Devuelve un objeto RespuestaPorDefectoAuditoria con data SeguimientoSocialDTO Paginada
     *
     * @param httpServletRequest request peticion.
     * @param bodyEncriptado     PaginacionRequest datos para obtener todos los seguimientos sociales.
     * @return RespuestaPorDefectoAuditoria<SeguimientoSocialDTO>
     */
    RespuestaPorDefectoAuditoria<PaginacionResponse<SeguimientoSocialDTO>> obtenerSeguimientosSociales(
            HttpServletRequest httpServletRequest,
            BodyEncriptado bodyEncriptado);

    /**
     * Elimina un seguimiento social del sistema
     *
     * @param httpServletRequest HttpServletRequest datos del request.
     * @param bodyEncriptado     PaginacionRequest datos del seguimiento social a eliminar.
     * @return RespuestaPorDefectoAuditoria<Boolean>
     */
    RespuestaPorDefectoAuditoria<Boolean> eliminarSeguimientoSocial(
            HttpServletRequest httpServletRequest,
            BodyEncriptado bodyEncriptado);

    /**
     * Devuelve un objeto RespuestaPorDefectoAuditoria con data SeguimientoSocialDTO si el seguimiento social se creó con éxito
     *
     * @param httpServletRequest request petición.
     * @param bodyEncriptado     objeto SeguimientoSocialDTO a crear.
     * @return RespuestaPorDefectoAuditoria<SeguimientoSocialDTO>
     */
    RespuestaPorDefectoAuditoria<SeguimientoSocialDTO> crearSeguimientoSocial(
            HttpServletRequest httpServletRequest,
            BodyEncriptado bodyEncriptado);

    /**
     * Sube un documento y lo asocia al registro respectivo de evaluación domiciliaria
     *
     * @param httpServletRequest Solicitud HTTP
     * @param bodyEncriptado     Cuerpo encriptado con los datos de relación
     * @param multipartFiles     Archivo a subir
     * @return RespuestaPorDefectoAuditoria<DocumentoDTO>
     */
    RespuestaPorDefectoAuditoria<Boolean> subirDocumentos(HttpServletRequest httpServletRequest,
                                                          BodyEncriptado bodyEncriptado,
                                                          MultipartFile[] multipartFiles);

    /**
     * Obtiene todos los documentos asociados al registro de evaluación domiciliaria
     *
     * @param httpServletRequest Solicitud HTTP
     * @param bodyEncriptado     Request con parámetros de paginación y búsqueda
     * @return RespuestaPorDefectoAuditoria<PaginacionResponse <DocumentoDTO>>
     */
    RespuestaPorDefectoAuditoria<PaginacionResponse<DocumentoDTO>> obtenerDocumentos(HttpServletRequest httpServletRequest,
                                                                                     BodyEncriptado bodyEncriptado);

}